Framework/Core/include/Framework/TMessageSerializer.h

```diff
@@ -13,41 +13,92 @@
 #include <fairmq/FairMQMessage.h>
 
 #include <TMessage.h>
-#include <TClonesArray.h>
+#include <TObjArray.h>
+#include <TList.h>
+#include <mutex>
+#include <TStreamerInfo.h>
+#include <memory>
 
 namespace o2 {
 namespace framework {
 
 class FairTMessage : public TMessage
 {
   public:
+    using TMessage::TMessage;
     FairTMessage(void* buf, Int_t len)
         : TMessage(buf, len)
     {
         ResetBit(kIsOwner);
     }
-};
 
-// helper function to clean up the object holding the data after it is transported.
-static void free_tmessage(void* /*data*/, void* hint)
-{
-    delete (TMessage*)hint;
-}
+    // helper function to clean up the object holding the data after it is transported.
+    static void free_fn(void* /*data*/, void* hint)
+    {
+      std::default_delete<FairTMessage> deleter;
+      deleter(static_cast<FairTMessage*>(hint));
+    }
+};
 
 struct TMessageSerializer
 {
-  void Serialize(FairMQMessage& msg, TObject* input)
+  using StreamerList = std::vector<TVirtualStreamerInfo*>;
+  using CompressionLevel = int;
+  enum class CacheStreamers { yes, no };
+
+  void Serialize(FairMQMessage& msg, TObject* input, CacheStreamers streamers = CacheStreamers::no,
+                 CompressionLevel compressionLevel = -1)
+  {
+    std::unique_ptr<FairTMessage> tm = std::make_unique<FairTMessage>(kMESS_OBJECT);
+
+    if (streamers==CacheStreamers::yes) {
+      tm->EnableSchemaEvolution(true);
+    }
+
+      if (compressionLevel >= 0) {
+      // if negative, skip to use ROOT default
+      tm->SetCompressionLevel(compressionLevel);
+    }
+
+    tm->WriteObject(static_cast<TObject*>(input));
+
+    if (streamers==CacheStreamers::yes) {
+      UpdateStreamers(*tm, sStreamers);
+    }
+
+    msg.Rebuild(tm->Buffer(), tm->BufferSize(), FairTMessage::free_fn, tm.get());
+    tm.release();
+  }
+
+  void Deserialize(const FairMQMessage& msg, std::unique_ptr<TObject>& output)
   {
-    TMessage* tm = new TMessage(kMESS_OBJECT);
-    tm->WriteObject(static_cast<TObject *>(input));
-    msg.Rebuild(tm->Buffer(), tm->BufferSize(), free_tmessage, tm);
+    // we know the message will not be modified by this,
+    // so const_cast should be OK here(IMHO).
+    FairTMessage tm(const_cast<FairMQMessage&>(msg).GetData(), const_cast<FairMQMessage&>(msg).GetSize());
+    output.reset(reinterpret_cast<TObject*>(tm.ReadObject(tm.GetClass())));
   }
 
-  void Deserialize(FairMQMessage& msg, TObject*& output)
+  // load the schema information from a message
+  void LoadSchema(const FairMQMessage& msg);
```

Comment:
function name starts lower case, also for other functions

Reply:
done